zookeeper-server/src/main/java/org/apache/zookeeper/server/persistence/FileTxnSnapLog.java

```diff
@@ -232,11 +240,19 @@ public long restore(DataTree dt, Map<Long, Integer> sessions, PlayBackListener l
         } else {
             trustEmptyDB = autoCreateDB;
         }
+
         if (-1L == deserializeResult) {
             /* this means that we couldn't find any snapshot, so we need to
              * initialize an empty database (reported in ZOOKEEPER-2325) */
             if (txnLog.getLastLoggedZxid() != -1) {
-                throw new IOException("No snapshot found, but there are log entries. " + "Something is broken!");
+                // ZOOKEEPER-3056: provides an escape hatch for users upgrading
+                // from old versions of zookeeper (3.4.x, pre 3.5.3).
+                if (!trustEmptySnapshot) {
+                    throw new IOException(EMPTY_SNAPSHOT_WARNING + "Something is broken!");
+                } else {
+                    LOG.warn(EMPTY_SNAPSHOT_WARNING + "This should only be allowed during upgrading.");
+                    trustEmptySnapshot = false;
+                }
             }
 
             if (trustEmptyDB) {
```

Comment:
If `trustEmptyDB` is not set does the code (1) load the transaction log and (2) create a snapshot? With the second question, can we have a scenario where clearing `trustEmptySnapshot` causes the Peer to shutdown because it cannot find a snapshot on a second pass?

Reply:
I thought about this, and I didn't get into `trustEmptyDB` because it seems create another problem for upgrading, which might require a separate discussion.

Basically if we don't set `trustEmptyDB`, then it's possible to have zookeeper server in a valid state with no snapshot files but only transaction log files similar to older version of ZooKeeper. Since we treat such case as invalid, yet we essentially allow such case happen in first place, this seems self contradictory and inconsistent. Always creating at least one snapshot sounds like a good solution here, but that require we always execute the code path when `trustEmptyDB ` (which, is the default case).

Is there any cases where we don't `trustEmptyDB ` ?

Reply:
One concrete example:
* Create a fresh zookeeper cluster with `zookeeper.db.autocreate` set to false.
* Start a fresh cluster, create a client and issue some transactions.
* The default snap count is 100000 so no snapshot will be created; `zookeeper.db.autocreate` is set to false so no snapshot will be created upon server start up.
* Now restart all zookeeper servers. We can't load because there is no snapshot. 

Does this sound a reasonable example to demonstrate the problem with `trustEmptyDB ` set to false?